File: clients/android/app/src/test/java/id/my/anciety/codever/client/MatrixNativeProjectionTest.kt
package id.my.anciety.codever.client

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertThrows
import org.junit.Test

class MatrixNativeProjectionTest {
    @Test
    fun `builds inventory from independent Gateway and session Room State`() {
        val projection = MatrixNativeProjection()
        val entity = sessionState("session-1", 2)
        assertNull(projection.applyRoomState(entity))
        val state = projection.applyRoomState(gatewayState(2))!!
        assertEquals("session-1", state.sessions().single().getValue("id").jsonPrimitive.content)
    }

    @Test
    fun `Gateway first exposes an empty view then accepts session state independently`() {
        val projection = MatrixNativeProjection()
        assertEquals(0, projection.applyRoomState(gatewayState(1))!!.sessions().size)
        val state = projection.applyRoomState(sessionState("session-1", 2))!!
        assertEquals("session-1", state.sessions().single().getValue("id").jsonPrimitive.content)
    }

    @Test
    fun `complete Room State batch publishes atomically and invalid batch rolls back`() {
        val projection = MatrixNativeProjection()
        val snapshot = projection.applyRoomStateBatch(listOf(
            gatewayState(2),
            sessionState("session-1", 2),
            sessionState("session-2", 3),
        ))!!
        assertEquals(setOf("session-1", "session-2"), snapshot.sessions().map {
            it.getValue("id").jsonPrimitive.content
        }.toSet())

        val before = projection.snapshot()
        assertThrows(IllegalArgumentException::class.java) {
            projection.applyRoomStateBatch(listOf(
                tombstone("session-1", 4),
                JsonObject(sessionState("broken", 5) + mapOf(
                    "state" to JsonPrimitive("invalid"),
                )),
            ))
        }
        assertEquals(before, projection.snapshot())
    }

    @Test
    fun `updates one session without waiting for a global inventory commit`() {
        val projection = MatrixNativeProjection()
        val before = sessionState("session-1", 2, "Before")
        val after = sessionState("session-1", 3, "After")
        projection.applyRoomState(before)
        projection.applyRoomState(gatewayState(2))
        projection.applyRoomState(after)
        assertEquals(
            "After",
            projection.snapshot()!!.sessions().single()
                .getValue("title").jsonPrimitive.content,
        )
    }

    @Test
    fun `timeline history advances revision but never creates inventory`() {
        val projection = MatrixNativeProjection()
        projection.applyRoomState(gatewayState(1))
        projection.applyTimeline(buildJsonObject {
            put("version", 2); put("kind", "session_root"); putRevision(4)
            put("session_id", "timeline-only")
        })
        assertEquals(0, projection.snapshot()!!.sessions().size)
        assertEquals("4", projection.snapshot()!!.getValue("revision").jsonPrimitive.content)
    }

    @Test
    fun `tombstone prevents stale state from resurrecting a deleted session`() {
        val projection = MatrixNativeProjection()
        val current = sessionState("session-1", 2)
        val deleted = tombstone("session-1", 4)
        projection.applyRoomState(current)
        projection.applyRoomState(gatewayState(2))
        projection.applyRoomState(deleted)
        projection.applyRoomState(sessionState("session-1", 3, "Stale"))
        assertEquals(0, projection.snapshot()!!.sessions().size)
    }

    @Test
    fun `new revision generation discards retired generation entities`() {
        val projection = MatrixNativeProjection()
        val old = sessionState("old", 2)
        projection.applyRoomState(old)
        projection.applyRoomState(gatewayState(2))
        val next = JsonObject(gatewayState(1) + mapOf(
            "revision_epoch" to JsonPrimitive("epoch-2"),
            "revision_epoch_generation" to JsonPrimitive(2),
        ))
        projection.applyRoomState(next)
        assertEquals(0, projection.snapshot()!!.sessions().size)
    }

    @Test
    fun `rejects incomplete or duplicate Gateway capabilities`() {
        val projection = MatrixNativeProjection()
        assertThrows(IllegalArgumentException::class.java) {
            projection.applyRoomState(JsonObject(gatewayState(1) + mapOf(
                "capabilities" to buildJsonObject { put("can_create_session", true) },
            )))
        }
        val duplicateModes = capabilities(permissionModeIds = listOf("default", "default"))
        assertThrows(IllegalArgumentException::class.java) {
            projection.applyRoomState(JsonObject(gatewayState(1) + mapOf(
                "capabilities" to duplicateModes,
            )))
        }
    }

    private fun gatewayState(version: Long) = buildJsonObject {
        put("version", 2); put("kind", "gateway_state")
        put("gateway_id", "gateway-1"); put("conversation_id", "conversation-1")
        putRevision(version); put("state_version", version)
        put("active_device_count", 1)
        put("workspace", buildJsonObject {
            put("project", buildJsonObject {
                put("id", "project-1"); put("name", "codever"); put("cwd", "/repo")
            })
            put("provider", "codex"); put("permission_mode", "default")
        })
        put("capabilities", capabilities())
        put("updated_at", version)
    }

    private fun capabilities(
        permissionModeIds: List<String> = listOf("default"),
    ) = buildJsonObject {
        put("models", JsonArray(emptyList()))
        put("permission_modes", JsonArray(permissionModeIds.map { id ->
            buildJsonObject { put("id", id); put("name", "Default") }
        }))
        put("can_create_session", true)
        put("can_select_session", false)
        put("can_archive_session", true)
        put("can_delete_session", true)
        put("session_extensions", JsonArray(emptyList()))
    }

    private fun sessionState(id: String, version: Long, title: String = "Work") = buildJsonObject {
        put("version", 2); put("kind", "session_state")
        put("gateway_id", "gateway-1"); put("conversation_id", "conversation-1")
        putRevision(version); put("state_version", version)
        put("session_id", id); put("state", "active")
        put("session", buildJsonObject {
            put("session_id", id); put("title", title); put("updated_at", version)
            put("archived", false); put("status", "idle")
            put("project", buildJsonObject {
                put("id", "project-1"); put("name", "codever"); put("cwd", "/repo")
            })
            put("provider", "codex"); put("extensions", JsonArray(emptyList()))
        })
        put("updated_at", version)
    }

    private fun tombstone(id: String, version: Long) = buildJsonObject {
        put("version", 2); put("kind", "session_state")
        put("gateway_id", "gateway-1"); put("conversation_id", "conversation-1")
        putRevision(version); put("state_version", version)
        put("session_id", id); put("state", "deleted"); put("updated_at", version)
    }

    private fun kotlinx.serialization.json.JsonObjectBuilder.putRevision(revision: Long) {
        put("revision", revision); put("revision_epoch", "epoch-1")
        put("revision_epoch_generation", 1)
    }

    private fun JsonObject.sessions() = getValue("sessions").jsonArray.map { it.jsonObject }
    private fun JsonObject.text(key: String) = getValue(key).jsonPrimitive.content
}

File: apps/pwa/app/matrixNativeProjection.test.ts
import { describe, expect, it } from "vitest";
import { MatrixNativeProjection } from "./matrixNativeProjection";

describe("MatrixNativeProjection", () => {
  it("rebuilds session inventory from roots and lifecycle deltas", () => {
    const projection = new MatrixNativeProjection();
    projection.apply(checkpoint());
    projection.apply({
      version: 2,
      kind: "session_root",
      ...revision(1),
      session_id: "s1",
      title: "Investigate sync",
      project: { id: "p1", name: "codever", cwd: "/repo" },
      created_at: 1,
      updated_at: 1,
      archived: false,
      status: "running",
      provider: "codex",
      permission_mode: "default",
      extensions: [],
    });
    const state = projection.snapshot();
    expect(state?.sessions).toMatchObject([{ id: "s1", status: "running" }]);
    expect(projection.apply({
      version: 2,
      kind: "session_lifecycle",
      ...revision(2),
      session_id: "s1",
      state: "archived",
      updated_at: 2,
    })?.sessions[0]?.status).toBe("archived");
    expect(projection.apply({
      version: 2,
      kind: "session_lifecycle",
      ...revision(3),
      session_id: "s1",
      state: "deleted",
      updated_at: 3,
    })?.sessions).toEqual([]);
    expect(projection.snapshot()?.revision).toBe(3);
  });

  it("bootstraps sessions from the latest checkpoint without historical roots", () => {
    const projection = new MatrixNativeProjection();
    const state = projection.apply({
      ...checkpoint(),
      sessions: [{
        session_id: "s-existing",
        title: "Existing work",
        updated_at: 7,
        archived: false,
        status: "idle" as const,
        project: { id: "p1", name: "codever", cwd: "/repo" },
        provider: "codex",
        extensions: [],
      }],
    });

    expect(state?.sessions).toMatchObject([{
      id: "s-existing",
      title: "Existing work",
      projectName: "codever",
    }]);
  });

  it("does not let a duplicate checkpoint erase newer session roots", () => {
    const projection = new MatrixNativeProjection();
    const initial = {
      ...checkpoint(),
      sessions: [{
        session_id: "s-existing",
        title: "Existing work",
        updated_at: 7,
        archived: false,
        status: "idle" as const,
        project: { id: "p1", name: "codever", cwd: "/repo" },
        provider: "codex",
        extensions: [],
      }],
    };
    projection.apply(initial);
    projection.apply({
      version: 2,
      kind: "session_root",
      ...revision(2),
      session_id: "s-new",
      title: "New work",
      project: { id: "p2", name: "another", cwd: "/another" },
      created_at: 8,
      updated_at: 8,
      archived: false,
      status: "idle",
      provider: "codex",
      permission_mode: "default",
      extensions: [],
    });

    const replayed = projection.apply(initial);

    expect(replayed?.sessions.map((session) => session.id).sort()).toEqual([
      "s-existing",
      "s-new",
    ]);
  });

  it("replays updates that arrive before their thread root", () => {
    const projection = new MatrixNativeProjection();
    projection.apply(checkpoint());
    projection.apply({
      version: 2,
      kind: "session_update",
      ...revision(2),
      session_id: "s1",
      updated_at: 3,
      title: "Recovered title",
      model: null,
    });
    projection.apply({
      version: 2,
      kind: "session_lifecycle",
      ...revision(3),
      session_id: "s1",
      state: "archived",
      updated_at: 4,
    });

    const state = projection.apply({
      version: 2,
      kind: "session_root",
      ...revision(1),
      session_id: "s1",
      title: "Initial title",
      project: { id: "p1", name: "codever", cwd: "/repo" },
      created_at: 1,
      updated_at: 1,
      archived: false,
      status: "idle",
      provider: "codex",
      model: "old-model",
      permission_mode: "default",
      extensions: [],
    });

    expect(state?.sessions).toMatchObject([{
      id: "s1",
      title: "Recovered title",
      status: "archived",
    }]);
    expect(state?.sessions[0]).not.toHaveProperty("model");
  });

  it("does not resurrect a session deleted before its root is fetched", () => {
    const projection = new MatrixNativeProjection();
    projection.apply(checkpoint());
    projection.apply({
      version: 2,
      kind: "session_lifecycle",
      ...revision(2),
      session_id: "s1",
      state: "deleted",
      updated_at: 4,
    });
    expect(projection.apply({
      version: 2,
      kind: "session_root",
      ...revision(1),
      session_id: "s1",
      title: "Deleted session",
      project: { id: "p1", name: "codever", cwd: "/repo" },
      created_at: 1,
      updated_at: 1,
      archived: false,
      status: "idle",
      provider: "codex",
      permission_mode: "default",
      extensions: [],
    })?.sessions).toEqual([]);
  });

  it("advances revision from a lightweight room timeline event", () => {
    const projection = new MatrixNativeProjection();
    projection.apply(checkpoint());
    const state = projection.apply({
      version: 2,
      kind: "gateway_revision",
      ...revision(4),
      gateway_id: "g1",
      conversation_id: "c1",
      updated_at: 4,
      source_command_id: "command-4",
    });
    expect(state?.revision).toBe(4);
    expect(state?.sessions).toEqual([]);
  });
});

function checkpoint() {
  return {
    version: 2,
    kind: "gateway_checkpoint",
    gateway_id: "g1",
    conversation_id: "c1",
    revision: 1,
    revision_epoch: "r1",
    revision_epoch_generation: 1,
    state_version: 1,
    active_device_count: 1,
    sessions: [],
    workspace: {
      project: { id: "p1", name: "codever", cwd: "/repo" },
      provider: "codex",
      permission_mode: "default",
    },
    capabilities: {
      models: [],
      permission_modes: [{ id: "default", name: "Default" }],
      can_create_session: true,
      can_select_session: false,
      can_archive_session: true,
      can_delete_session: true,
      session_extensions: [],
    },
    updated_at: 1,
  } as const;
}

function revision(value: number) {
  return {
    revision: value,
    revision_epoch: "r1",
    revision_epoch_generation: 1,
  } as const;
}
